File: app/frontend/src/features/announcements/domain/announcement.ts
/**
 * Announcement Domain - お知らせドメインモデル
 *
 * お知らせ機能のドメインエンティティと判定関数。
 *
 * ## 将来のバックエンド連携について（Outbox パターン推奨）
 *
 * 【DB設計】
 * - announcements テーブルに notification_plan を JSON カラムで保存
 * - publish 時に notification_outbox テーブルにレコードを積む
 *   - columns: id, announcement_id, channel, status, sent_at, error_message, created_at
 *
 * 【Worker 処理】
 * - notification_outbox を定期的にポーリング
 * - status='pending' のレコードを処理
 * - email/line を送信し、status='sent'/'failed' に更新
 *
 * 【Clean Architecture】
 * - application: NotificationDispatcherPort（インターフェース）
 * - infra: EmailAdapter / LineAdapter（実装）
 *
 * TODO: バックエンド実装時にこのコメントを参照してください。
 */

/**
 * 重要度レベル
 */
export type AnnouncementSeverity = "info" | "warn" | "critical";

/**
 * 対象オーディエンス
 * - 'all': 全ユーザー
 * - 'internal': 社内ユーザー（今は all と同じ扱い）
 * - 'site:narita': 成田拠点のユーザー
 * - 'site:shinkiba': 新木場拠点のユーザー
 */
export type Audience = "all" | "internal" | "site:narita" | "site:shinkiba";

/**
 * 添付ファイル
 */
export interface Attachment {
  /** 表示ラベル */
  label: string;
  /** URL */
  url: string;
  /** ファイル種別 */
  kind?: "pdf" | "link";
}

/**
 * 通知チャネル
 */
export type NotificationChannel = "inApp" | "email" | "line";

/**
 * 通知設定
 *
 * 将来のメール/LINE連携のための設定情報。
 * 今回は表示のみ（送信は実装しない）。
 */
export interface NotificationPlan {
  /** 配信チャネル */
  channels: NotificationChannel[];
  /** 公開開始時に送信するか（将来実装） */
  sendOnPublish: boolean;
  /** 明示的なスケジュール日時（ISO8601、将来実装） */
  scheduledAt?: string | null;
  /** テンプレートヒント（将来のテンプレート切替用） */
  templateHint?: string | null;
}

/**
 * お知らせエンティティ
 */
export interface Announcement {
  /** 一意識別子 */
  id: string;
  /** タイトル */
  title: string;
  /** 本文（Markdown形式） */
  bodyMd: string;
  /** 重要度 */
  severity: AnnouncementSeverity;
  /** タグ（任意、最大2〜3個表示推奨） */
  tags?: string[];
  /** 公開開始日時（ISO8601） */
  publishFrom: string;
  /** 公開終了日時（ISO8601、null=無期限） */
  publishTo: string | null;
  /** 対象オーディエンス */
  audience: Audience;
  /** 添付ファイル */
  attachments?: Attachment[];
  /** 通知設定（無ければ inApp のみ相当） */
  notification?: NotificationPlan;
}

/**
 * お知らせがアクティブ（公開期間内）かどうかを判定
 *
 * @param announcement - 判定対象のお知らせ
 * @param now - 現在日時（デフォルト: 現在時刻）
 * @returns アクティブならtrue
 */
export function isAnnouncementActive(
  announcement: Announcement,
  now: Date = new Date(),
): boolean {
  const publishFrom = new Date(announcement.publishFrom);
  const publishTo = announcement.publishTo
    ? new Date(announcement.publishTo)
    : null;

  return publishFrom <= now && (publishTo === null || now <= publishTo);
}

/**
 * お知らせが指定オーディエンスに表示可能かどうかを判定
 *
 * @param announcement - 判定対象のお知らせ
 * @param currentAudience - 現在のユーザーオーディエンス
 * @returns 表示可能ならtrue
 *
 * TODO: 将来的にはユーザープロファイルから currentAudience を取得する
 */
export function isVisibleForAudience(
  announcement: Announcement,
  currentAudience: Audience,
): boolean {
  const { audience } = announcement;

  // 'all' と 'internal' は全員に表示（今は同じ扱い）
  if (audience === "all" || audience === "internal") {
    return true;
  }

  // サイト指定の場合は一致するときのみ表示
  return audience === currentAudience;
}

/**
 * バナー表示対象かどうかを判定
 *
 * 重要度が warn または critical のお知らせがバナー対象
 *
 * @param announcement - 判定対象のお知らせ
 * @returns バナー対象ならtrue
 */
export function isBannerTarget(announcement: Announcement): boolean {
  return (
    announcement.severity === "warn" || announcement.severity === "critical"
  );
}

File: app/backend/core_api/app/core/domain/announcement.py
"""
Announcement domain entities.
お知らせ（アナウンスメント）のドメインエンティティ
"""
from datetime import datetime
from typing import Optional, List, Literal
from pydantic import BaseModel, Field


AnnouncementSeverity = Literal["info", "warn", "critical"]
Audience = Literal["all", "internal", "site:narita", "site:shinkiba"]


class Attachment(BaseModel):
    """添付ファイル情報"""
    label: str = Field(..., description="表示ラベル")
    url: str = Field(..., description="ファイルURL")


class NotificationPlan(BaseModel):
    """通知設定"""
    email: bool = Field(False, description="メール通知")
    in_app: bool = Field(True, description="アプリ内通知")


class Announcement(BaseModel):
    """
    お知らせデータ (app.announcements)
    
    Fields:
        id: お知らせID (PK)
        title: タイトル
        body_md: 本文（Markdown形式）
        severity: 重要度 ('info' | 'warn' | 'critical')
        tags: タグ配列
        publish_from: 公開開始日時
        publish_to: 公開終了日時（NULL=無期限）
        audience: 対象 ('all' | 'internal' | 'site:narita' | 'site:shinkiba')
        attachments: 添付ファイル配列
        notification_plan: 通知設定
        created_at: 作成日時
        updated_at: 更新日時
    """
    id: int = Field(..., description="お知らせID")
    title: str = Field(..., description="タイトル")
    body_md: str = Field(..., description="本文（Markdown形式）")
    severity: AnnouncementSeverity = Field("info", description="重要度")
    tags: List[str] = Field(default_factory=list, description="タグ配列")
    publish_from: datetime = Field(..., description="公開開始日時")
    publish_to: Optional[datetime] = Field(None, description="公開終了日時")
    audience: Audience = Field("all", description="対象")
    attachments: List[Attachment] = Field(default_factory=list, description="添付ファイル配列")
    notification_plan: Optional[NotificationPlan] = Field(None, description="通知設定")
    created_at: datetime = Field(..., description="作成日時")
    updated_at: datetime = Field(..., description="更新日時")

    class Config:
        from_attributes = True


class AnnouncementUserState(BaseModel):
    """
    ユーザーごとのお知らせ既読・確認状態 (app.announcement_user_states)
    
    Fields:
        id: 状態ID (PK)
        user_id: ユーザー識別子
        announcement_id: お知らせID (FK)
        read_at: 既読日時（NULL=未読）
        ack_at: 確認日時（NULL=未確認、criticalお知らせ用）
    """
    id: int = Field(..., description="状態ID")
    user_id: str = Field(..., description="ユーザー識別子")
    announcement_id: int = Field(..., description="お知らせID")
    read_at: Optional[datetime] = Field(None, description="既読日時")
    ack_at: Optional[datetime] = Field(None, description="確認日時")

    class Config:
        from_attributes = True


class AnnouncementWithState(BaseModel):
    """
    ユーザー状態を含むお知らせデータ
    APIレスポンス用
    """
    announcement: Announcement
    read_at: Optional[datetime] = None
    ack_at: Optional[datetime] = None

    class Config:
        from_attributes = True
